src/libsyntax_ext/proc_macro_registrar.rs

```diff
@@ -232,6 +189,112 @@ impl<'a> Visitor<'a> for CollectCustomDerives<'a> {
+        // it.
+        let is_fn = match item.node {
+            ast::ItemKind::Fn(..) => true,
+            _ => false,
+        };
+
+        let mut found_attr: Option<&'a ast::Attribute> = None;
+
+        for attr in &item.attrs {
+            if attr.check_name("proc_macro_derive") || attr.check_name("proc_macro_attribute"){
+                if let Some(prev_attr) = found_attr {
+                    self.handler.span_err(attr.span(), "Only one `#[proc_macro_derive]` attribute \
+                                                        is allowed on any given function");
+                    self.handler.span_err(prev_attr.span(), "Previous procedural macro attribute \
+                                                                here");
+                    return;
+                }
+
+                found_attr = Some(attr);
+            }
+        }
+
+        let attr = match found_attr {
+            None => {
+                self.check_not_pub_in_root(&item.vis, item.span);
+                return visit::walk_item(self, item);
+            },
+            Some(attr) => attr,
+        };
+
+        if !is_fn {
+            let msg = if attr.check_name("proc_macro_derive") {
+                "the `#[proc_macro_derive]` attribute may only be used on bare functions"
+            } else if attr.check_name("proc_macro_attribute") {
+                "the `#[proc_macro_attribute]` attribute may only be used on bare functions"
+            } else {
+                unreachable!("`found_attr` was `Some` but not one of \
+                    `#[proc_macro_derive]` or `#[proc_macro_attribute]`: {:?}", attr);
+            };
+
+            self.handler.span_err(attr.span(), msg);
+            return;
+        }
+
+        if self.is_test_crate {
+            return;
+        }
+
+        if !self.is_proc_macro_crate {
+            let msg = if attr.check_name("proc_macro_derive") {
+                "the `#[proc_macro_derive]` attribute is \
+                    only usable with crates of the `proc-macro` \
+                    crate type"
+            } else if attr.check_name("proc_macro_attribute") {
+                "the `#[proc_macro_attribute]` attribute is \
+                    only usable with crates of the `proc-macro` \
+                    crate type"
+            } else {
+                unreachable!("`found_attr` was `Some` but not one of \
+                    `#[proc_macro_derive]` or `#[proc_macro_attribute]`: {:?}", attr);
+            };
+
+            self.handler.span_err(attr.span(), msg);
+            return;
+        }
+
+        if attr.check_name("proc_macro_derive") {
+            self.collect_custom_derive(item, attr);
+        } else if attr.check_name("proc_macro_attribute") {
+            self.collect_attr_proc_macro(item, attr);
+        } else {
+            unreachable!("`found_attr` was `Some` but not one of \
+                    `#[proc_macro_derive]` or `#[proc_macro_attribute]`: {:?}", attr);
```

Comment:
IMO, it's easy enough to see that [this](https://github.com/rust-lang/rust/pull/38842/files#diff-87e54b77e8920340f8132507d88e6798R242) is the only assignment of `found_attrs` (proving that this is unreachable) that the detailed message in the `unreachable!()` isn't warrented.